src/eventHandlers/InvocationHandler.ts

```diff
@@ -117,7 +120,8 @@ export class InvocationHandler extends EventHandler<'invocationRequest', 'invoca
             }
 
             const postInvocContext: PostInvocationContext = {
-                hookData,
+                hookData: preInvocContext.hookData,

```

Comment:
I don't think you meant to change behavior here, but you actually did. I would suggest changing this back to this:
```suggestion
                hookData: invocationHookData,
```

The behavior in question is "Should we let a hook change `hookData` to a different object, essentially erasing any existing state?" Aka in this example code should it print 'world' at the end or not:
```
// pre invoc hook 1
preInvocContext.hookData.hello = 'world';

// pre invoc hook 2
preInvocContext.hookData = {};

// post invoc hook 1
console.log(postInvocContext.hookData.hello);
```

I think the answer should be no, mainly because I expect these hooks can come from multiple sources that have no idea about each other and we should do our best to prevent one source from messing up another (e.g. one source is the user's app, another source is the "applicationinsights" npm package).

We are a bit inconsistent in how we handle this situation before your PR. We let users overwrite hookData within the same hook type, but not across hook types. To fix that, I would suggest we modify `WorkerChannel.executeHooks` to have the following code. You can either include it in your PR or let me do it as a separate PR. At a minimum, your PR should maintain existing behavior
```typescript
for (const callback of callbacks) {
    const oldHookData = context.hookData;
    const oldAppHookData = context.appHookData;
    await callback(context);
    if (context.hookData !== oldHookData || context.appHookData !== oldAppHookData) {
        context.hookData = oldHookData;
        context.appHookData = oldAppHookData;
        this.log({
            message:
                'WARNING: Ignoring changes to hook data. You may modify the hook data object, but cannot replace it.',
            level: LogLevel.Warning,
            logCategory: LogCategory.System,
            invocationId,
            category: msgCategory,
        });
    }
}
```

Reply:
This actually was an intentional change, as I thought the original behavior was an oversight. In my opinion, I thought that setting the `hookData` state through something like:

```TS
preInvocContext.hookData = {test: 'hello'};
```

is an intuitive way of setting this `hookData` object (see the unit tests I wrote for example), and I would be confused why setting it that way, I'm not able to see the new state in my post-invocation hook. I do see your point though about not wanting different sources to influence each other, but I think that is somewhat inevitable and should be expected from these hooks, since if we _don't_ allow overriding, then user hooks will be forced to see state from app insights, for example. What if they both try to set a value of the same name and they don't get the behavior they expect? Point is, some interference between the two is inevitable, regardless of which way we go.

Another suggestion would be, we don't allow `preInvocationContext.hookData = {}` to _override_ existing state, but we do allow it to _add_ new state. For example:

```TS
// pre invoc hook 1
preInvocContext.hookData.hello = 'world';

// pre invoc hook 2
preInvocContext.hookData = { test: 'hello' };

// post invoc hook 1
console.log(postInvocContext.hookData);
```
the value printed would be essentially the same behavior as `Object.assign`: 

```TS
{
    hello: 'world',
    test: 'hello'
}
```

That being said, regardless of which behavior we take, I think we should have consistent behavior between `appHookData` and `hookData`, and also consistent behavior for `hookData` for the different scopes. For example, if `preInvocationContext.hookData = {}` is not honored, then `preInvocationContext.appHookData = {}` should also not be honored.

Reply:
> In my opinion, I thought that setting the hookData state through something like is an intuitive way of setting this hookData object (see the unit tests I wrote for example)

I disagree that it's intuitive because it inherently only works for one hook. Ignoring multiple sources, if you have multiple hooks in your own app this would overwrite your other hook's data every time. `hookData` is like a shared photo album with your family. People can add and remove photos, but they can't create a new album each time because that would destroy everyone else's contributions.

> some interference between the two is inevitable, regardless of which way we go.

Yes and that's by design. In order to share data, we have to let users interfere with each other to some extent. I would recommend the app insights team put all of their data in a nested object with a very unique key (`microsoft-applicationinsights`?) to prevent accidental modifications. The programming model package can also help make it easier to generate unique keys, but I don't want that kind of logic in the core module.

> That being said, regardless of which behavior we take, I think we should have consistent behavior between appHookData and hookData, and also consistent behavior for hookData for the different scopes.

I agree with this - but that's not the goal of this PR. This has already taken longer than I would like for any PR, meaning it was fixing too many things at once (not that the code itself is bad). Keep the scope small, maintain existing behavior if it's unrelated to the title of the PR, and we should be able to get every PR merged within a week or two IMO.